akka-typed/src/main/scala/akka/typed/internal/receptionist/ReceptionistImpl.scala

```diff
@@ -0,0 +1,136 @@
+/**
+ * Copyright (C) 2009-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.typed.internal.receptionist
+
+import akka.annotation.InternalApi
+import akka.typed.ActorRef
+import akka.typed.Behavior
+import akka.typed.Terminated
+import akka.typed.receptionist.Receptionist._
+import akka.typed.scaladsl.Actor
+import akka.typed.scaladsl.Actor.immutable
+import akka.typed.scaladsl.Actor.same
+import akka.typed.scaladsl.ActorContext
+import akka.util.TypedMultiMap
+
+import scala.reflect.ClassTag
+
+/** Internal API */
+@InternalApi
+private[typed] object ReceptionistImpl {
+  // FIXME: make sure to provide serializer
+  case class DefaultServiceKey[T](id: String)(implicit tTag: ClassTag[T]) extends ServiceKey[T] {
+    override def toString: String = s"ServiceKey[$tTag]($id)"
+  }
+
+  /**
+   * Interface to allow plugging of external service discovery infrastructure in to the existing receptionist API.
+   *
+   * FIXME: find better name
+   */
+  trait ExternalInterface {
+    def onRegister[T](key: ServiceKey[T], address: ActorRef[T]): Unit
+    def onUnregister[T](key: ServiceKey[T], address: ActorRef[T]): Unit
+  }
+  object LocalExternalInterface extends ExternalInterface {
+    def onRegister[T](key: ServiceKey[T], address: ActorRef[T]): Unit = ()
+    def onUnregister[T](key: ServiceKey[T], address: ActorRef[T]): Unit = ()
+  }
+
+  val onlyLocalBehavior: Behavior[Command] = init(_ ⇒ LocalExternalInterface)
+
+  type KV[K <: AbstractServiceKey] = ActorRef[K#Protocol]
+  type ServiceMap = TypedMultiMap[AbstractServiceKey, KV]
+  object ServiceMap {
+    val empty: ServiceMap = TypedMultiMap.empty[AbstractServiceKey, KV]
+  }
+
+  sealed abstract class ReceptionistInternalCommand extends InternalCommand
+  final case class RegisteredActorTerminated[T](key: ServiceKey[T], address: ActorRef[T]) extends ReceptionistInternalCommand
+  final case class SubscriberTerminated[T](key: ServiceKey[T], address: ActorRef[Listing[T]]) extends ReceptionistInternalCommand
+  final case class RegistrationsChangedExternally(changes: ServiceMap) extends ReceptionistInternalCommand
+
+  type SubscriptionsKV[K <: AbstractServiceKey] = ActorRef[Listing[K#Protocol]]
+  type SubscriptionMap = TypedMultiMap[AbstractServiceKey, SubscriptionsKV]
+
+  private[receptionist] def init(externalInterfaceFactory: ActorContext[AllCommands] ⇒ ExternalInterface): Behavior[Command] =
```

Comment:
I wonder what you think about that `ExternalInterface` mechanism to plugin in the cluster logic without duplicating all of the rest. It doesn't feel quite right that there's no better, composable way of adding replication to the local logic. On the other hand, adding those hooks makes it quite explicit where the plugging points are.

In the end, I guess it doesn't matter for now how it is implemented exactly if it allows us to make it work first.

Reply:
I think that is looking good